File: cameo-api-cpp/src/coms/impl/ResponderImpl.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "ResponderImpl.h"

#include "RequestImpl.h"
#include "Application.h"
#include "Serializer.h"
#include "JSON.h"
#include "../../base/impl/RequestSocketImpl.h"
#include "../../base/impl/ContextImpl.h"
#include "../../base/Messages.h"
#include <sstream>

namespace cameo {
namespace coms {

const std::string ResponderImpl::RESPONDER_PREFIX = "rep.";

ResponderImpl::ResponderImpl(int responderPort, const std::string& name) :
	m_responderPort(responderPort),
	m_name(name),
	m_canceled(false) {

	// create a socket REP
	ContextImpl* contextImpl = dynamic_cast<ContextImpl *>(application::This::getCom().getContext());
	m_responder.reset(new zmq::socket_t(contextImpl->m_context, ZMQ_REP));
	std::stringstream repEndpoint;
	repEndpoint << "tcp://*:" << m_responderPort;

	m_responder->bind(repEndpoint.str().c_str());
}

ResponderImpl::~ResponderImpl() {
	terminate();
}

void ResponderImpl::cancel() {

	json::StringObject request;
	request.pushKey(message::TYPE);
	request.pushInt(message::CANCEL);

	// Create a request socket.
	std::unique_ptr<RequestSocketImpl> requestSocket = application::This::getCom().createRequestSocket(application::This::getEndpoint().withPort(m_responderPort).toString());
	requestSocket->requestJSON(request.toString());
}

WaitingImpl * ResponderImpl::waiting() {
	return new GenericWaitingImpl(std::bind(&ResponderImpl::cancel, this));
}

std::unique_ptr<RequestImpl> ResponderImpl::receive() {

	std::unique_ptr<zmq::message_t> message(new zmq::message_t);
	m_responder->recv(message.get(), 0);

	// Get the JSON request.
	json::Object request;
	json::parse(request, message.get());

	int type = request[message::TYPE].GetInt();

	// Create the reply
	std::unique_ptr<zmq::message_t> reply;
	std::unique_ptr<RequestImpl> result;

	if (type == message::REQUEST) {

		std::string name = request[message::Request::APPLICATION_NAME].GetString();
		int id = request[message::Request::APPLICATION_ID].GetInt();
		std::string serverUrl = request[message::Request::SERVER_URL].GetString();
		int serverPort = request[message::Request::SERVER_PORT].GetInt();
		int requesterPort = request[message::Request::REQUESTER_PORT].GetInt();

		// Get the second part for the message.
		message.reset(new zmq::message_t);
		m_responder->recv(message.get(), 0);
		std::string message1(message->data<char>(), message->size());

		// Create the request.
		result = std::unique_ptr<RequestImpl>(new RequestImpl(name,
				id,
				message1,
				serverUrl,
				serverPort,
				requesterPort));

		// Set message 2 if it exists.
		if (message->more()) {
			message.reset(new zmq::message_t);
			m_responder->recv(message.get(), 0);
			result->m_message2 = std::string(message->data<char>(), message->size());
		}
		reply.reset(processRequest());
	}
	else if (type == message::CANCEL) {
		m_canceled = true;
		reply.reset(processCancelResponder());
	}
	else {
		reply.reset(processUnknownRequest());
	}

	// send to the client
	if (reply != nullptr) {
		m_responder->send(*reply);
	}

	return result;
}

zmq::message_t * ResponderImpl::processRequest() {

	std::string result = createRequestResponse(0, "OK");

	zmq::message_t * reply = new zmq::message_t(result.length());
	memcpy(reply->data(), result.c_str(), result.length());

	return reply;
}

zmq::message_t * ResponderImpl::processCancelResponder() {

	std::string result = createRequestResponse(0, "OK");

	zmq::message_t * reply = new zmq::message_t(result.length());
	memcpy(reply->data(), result.c_str(), result.length());

	return reply;
}

zmq::message_t * ResponderImpl::processUnknownRequest() {

	std::string result = createRequestResponse(-1, "Unknown request");

	zmq::message_t * reply = new zmq::message_t(result.length());
	memcpy(reply->data(), result.c_str(), result.length());

	return reply;
}

void ResponderImpl::terminate() {

	if (m_responder.get() != nullptr) {
		m_responder.reset(nullptr);

		application::This::getCom().removePort(RESPONDER_PREFIX + m_name);
	}
}

}
}

File: cameo-api/src/main/java/fr/ill/ics/cameo/coms/impl/ResponderImpl.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package fr.ill.ics.cameo.coms.impl;

import org.json.simple.JSONObject;

import fr.ill.ics.cameo.Zmq;
import fr.ill.ics.cameo.base.Application.This;
import fr.ill.ics.cameo.base.RequestSocket;
import fr.ill.ics.cameo.base.impl.zmq.ContextZmq;
import fr.ill.ics.cameo.messages.JSON;
import fr.ill.ics.cameo.messages.Messages;
import fr.ill.ics.cameo.strings.Endpoint;

public class ResponderImpl {

	public static final String RESPONDER_PREFIX = "rep.";

	private int responderPort;
	private String name;
	
	private Zmq.Context context;
	private Zmq.Socket responder;
	
	private boolean ended = false;
	private boolean canceled = false;
	private ResponderWaitingImpl waiting = new ResponderWaitingImpl(this);
	
	public ResponderImpl(int responderPort, String name) {
		this.responderPort = responderPort;
		this.name = name;
		this.context = ((ContextZmq)This.getCom().getContext()).getContext();

		// create a socket REP
		responder = context.createSocket(Zmq.REP);
		responder.bind("tcp://*:" + responderPort);
		
		waiting.add();
	}
	
	public String getName() {
		return name;
	}

	public RequestImpl receive() {
		
		Zmq.Msg message = null;
		Zmq.Msg reply = null;
		
		try {
			message = Zmq.Msg.recvMsg(responder);

			if (message == null) {
				ended = true;
				
				return null;
			}
			
			// Get the JSON request object.
			JSONObject request = This.getCom().parse(message.getFirstData());
			
			// Get the type.
			long type = JSON.getLong(request, Messages.TYPE);
			
			if (type == Messages.REQUEST) {

				String name = JSON.getString(request, Messages.Request.APPLICATION_NAME);
				int id = JSON.getInt(request, Messages.Request.APPLICATION_ID);
				String serverUrl = JSON.getString(request, Messages.Request.SERVER_URL);
				int serverPort = JSON.getInt(request, Messages.Request.SERVER_PORT);
				int requesterPort = JSON.getInt(request, Messages.Request.REQUESTER_PORT);
				
				byte[][] data = message.getAllData();
				
				byte[] message1 = data[1];
				
				// Create the request implementation.
				RequestImpl impl = new RequestImpl(name, 
						id, 
						message1, 
						serverUrl,
						serverPort,
						requesterPort);
				
				// Set the optional message 2.
				if (data.length > 2) {
					impl.setMessage2(data[2]);
				}
				
				return impl;
			}
			else if (type == Messages.CANCEL) {
				canceled = true;
				
				return null;
			}
		}
		finally {
			if (message != null) {
				message.destroy();
			}	

			// Send to the requester
			reply = responseToRequest();
			reply.send(responder);
			
			if (reply != null) {
				reply.destroy();
			}
		}
			
		return null;
	}
	
	public void cancel() {
		Endpoint endpoint = This.getEndpoint().withPort(responderPort);

		JSONObject request = new JSONObject();
		request.put(Messages.TYPE, Messages.CANCEL);
		
		// Create the request socket. We can create it here because it should be called only once.
		RequestSocket requestSocket = This.getCom().createRequestSocket(endpoint.toString());
		requestSocket.requestJSON(request);
		
		// Terminate the socket.
		requestSocket.terminate();
	}

	private Zmq.Msg responseToRequest() {
		
		Zmq.Msg message = new Zmq.Msg();
		message.add(Messages.serialize(Messages.createRequestResponse(0, "OK")));
		
		return message;
	}
	
	public boolean isEnded() {
		return ended;
	}
	
	public boolean isCanceled() {
		return canceled;
	}
	
	public void terminate() {

		waiting.remove();
		context.destroySocket(responder);
		
		try {
			This.getCom().removePort(RESPONDER_PREFIX + name);
			
		} catch (Exception e) {
			System.err.println("Cannot terminate responder: " + e.getMessage());
		}
	}
	
	@Override
	public String toString() {
		return RESPONDER_PREFIX + name + ":" + This.getName() + "." + This.getId() + "@" + This.getEndpoint();
	}
	
}